tests/integration/exec_test.py

```diff
@@ -88,6 +94,67 @@ def test_exec_command_streaming(self):
             res += chunk
         self.assertEqual(res, b'hello\nworld\n')
 
+    def test_exec_start_socket(self):
+        if not helpers.exec_driver_is_native():
+            pytest.skip('Exec driver not native')
+
+        container = self.client.create_container(BUSYBOX, 'cat',
+                                                 detach=True, stdin_open=True)
+        id = container['Id']
+        self.client.start(id)
+        self.tmp_containers.append(id)
+
+        line = 'yay, interactive exec!'
+        # `echo` appends CRLF, `printf` doesn't
+        exec_id = self.client.exec_create(id, ['printf', line], tty=True)
+        self.assertIn('Id', exec_id)
+
+        socket = self.client.exec_start(exec_id, socket=True)
+
+        select.select([socket], [], [])
+
+        recoverable_errors = (errno.EINTR, errno.EDEADLK, errno.EWOULDBLOCK)
+
+        def read(n=4096):
+            """Code stolen from dockerpty to read the socket"""
+            try:
+                if hasattr(socket, 'recv'):
+                    return socket.recv(n)
+                return os.read(socket.fileno(), n)
+            except EnvironmentError as e:
+                if e.errno not in recoverable_errors:
+                    raise
+
+        def next_packet_size():
+            """Code stolen from dockerpty to get the next packet size"""
+            data = six.binary_type()
+            while len(data) < 8:
+                next_data = read(8 - len(data))
+                if not next_data:
+                    return 0
+                data = data + next_data
+
+            if data is None:
+                return 0
+
+            if len(data) == 8:
+                _, actual = struct.unpack('>BxxxL', data)
+                return actual
+
+        try:
+            next_size = next_packet_size()
+            self.assertEqual(next_size, len(line))
+
+            data = six.binary_type()
+            while len(data) < next_size:
+                next_data = read(next_size - len(data))
+                if not next_data:
+                    assert False, "Failed trying to read in the dataz"
+                data += next_data
+            self.assertEqual(data.decode('utf-8'), "{0}".format(line))
+        finally:
+            socket.close()
```

Comment:
This `try/finally` could alternatively be done with a `self.addCleanup(socket.close)` line right after calling `exec_start`, which might be a bit more readable, as well as catching any errors raised by `select` (assuming it could raise any).